zerver/tests/test_queue_worker.py

```diff
@@ -214,32 +217,82 @@ def cancel(self) -> None:
-            ),
-        ]
+                rcpt_to=stream_to_address
+            )
+        ] * 3
         for element in data:
             fake_client.queue.append(('email_mirror', element))
 
-        with patch('zerver.worker.queue_processors.mirror_email'):
+        with patch('zerver.worker.queue_processors.mirror_email') as mirror_email:
             with simulated_queue_client(lambda: fake_client):
                 worker = queue_processors.MirrorWorker()
                 worker.setup()
                 worker.start()
 
+            self.assertEqual(mirror_email.call_count, 3)
+
+    def test_mirror_worker_rate_limiting(self) -> None:
+        fake_client = self.FakeClient()
+        realm = get_realm('zulip')
+        clear_history(RateLimitedRealmMirror(realm))
+        stream = get_stream('Denmark', realm)
+        stream_to_address = encode_email_address(stream)
+        data = [
+            dict(
+                message=u'\xf3test',
+                time=time.time(),
+                rcpt_to=stream_to_address
+            )
+        ] * 5
+        for element in data:
+            fake_client.queue.append(('email_mirror', element))
+
+        with patch('zerver.worker.queue_processors.mirror_email') as mirror_email:
+            with self.settings(RATE_LIMITING_MIRROR_REALM_RULES=[(10, 2)]):
+                with simulated_queue_client(lambda: fake_client):
+                    start_time = time.time()
+
+                    worker = queue_processors.MirrorWorker()
+                    worker.setup()
+                    worker.start()
+                    self.assertEqual(mirror_email.call_count, 2)
+
+                    # If a new message is sent into the stream mirror, it will get rejected:
+                    fake_client.queue.append(('email_mirror', data[0]))
+                    worker.start()
+                    self.assertEqual(mirror_email.call_count, 2)
+
+                    # However, missed message emails don't get rate limited:
+                    with self.settings(EMAIL_GATEWAY_PATTERN="%s@example.com"):
+                        address = 'mm' + ('x' * 32) + '@example.com'
+                        event = dict(
+                            message=u'\xf3test',
+                            time=time.time(),
+                            rcpt_to=address
+                        )
+                        fake_client.queue.append(('email_mirror', event))
+                        worker.start()
+                        self.assertEqual(mirror_email.call_count, 3)
+
+                    # After some times passes, emails get accepted again:
+                    with patch('time.time', return_value=(start_time + 11.0)):
+                        fake_client.queue.append(('email_mirror', data[0]))
+                        worker.start()
+                        self.assertEqual(mirror_email.call_count, 4)
+
+                        # Cover RateLimiterLockingException:
```

Comment:
I would say instead "If RateLimiterLockingException is thrown, we rate-limit the new message".